### PR TITLE
onRangeChange prop

### PR DESCRIPTION
**Pull request by jhockett**
Added an onRangeChange prop callback that fires when either the view or date range is changed. Returns a date range.

### CODE REVIEW

**Comment by jquense** [comment]
hey there, thanks for the PR. I'm a bit on the fence about this. I don't really want to add another api to the views, since it'll break or require changes from anyone that has a custom one. Static properties don't compose well either :?

**Comment by jhockett** [comment]
This functionality is absolutely necessary for the project I'm working on, and I think the knowing which dates are currently visible should be a fundamental feature of any calendar. 

Is there a way to make it more backwards compatible friendly?
As far as breaking changes for certain users, I agree, it sucks.
https://github.com/intljusticemission/react-big-calendar/issues/565

"Static properties don't compose well either :?"
I tried to follow the style of rbc as best I could, but I'm newer to React so I'm not sure what's wrong with what I did so please elaborate.

**Comment by tobiasandersen** [comment]
I get the concerns of @jquense, but I really like this feature since it's super useful for paging. @jquense what if we make it so that it doesn't break for views that hasn't implemented a range function? Could just skip calling `onRangeChange` if it's missing, and maybe show some warning in the console? 

Rather than having static methods, would you prefer having a `getRange` method on the instance instead?

**Comment by jquense** [comment]
I think the static methods are better than instance ones but that's a good idea  about not breaking for existing views at least :thumbsup:

**Comment by jhockett** [comment]
Sorry for the slow reply, I've been on holiday.

So the easiest thing for me to change would to just add a check to see if the view is custom at all.
`      if(!(view in VIEWS)) {
        console.warn('onRangeChange prop not supported for custom views')
        return
      }
`

I'd like to support custom views that support the api, my issue though is I don't know how to check to see if their view supports the api. With the default views, I can use `VIEWS[this.props.view]` to access the component, but I am not sure how to access a custom view component otherwise.

Side note: @tobiasandersen mentioned adding a warning in the console, but the linter will not allow it. How should I proceed on the matter?

**Comment by ehahn9** [comment]
+1 for getRange into view instance or add a viewInfo payload to the onView event?

fullCalendar has the ability to ask a view for its start/end range: https://fullcalendar.io/docs/views/View_Object/

Why is this imporant? Anyone fetching events from a server (we front-end gcal and o365 calendar) needs to know what event ranges to ask for (both of these support querying events over time ranges to limit the returned payload size).

As of now, we have external logic to deduce the range from the view name and date, but that's not a great encapsulation and requires us to keep up with any custom views added - which sucks bigtime, but certainly workable!

Just $0.02 on this.

File: src/Calendar.js [hunk 1]
```diff
@@ -679,6 +685,19 @@ class Calendar extends React.Component {
     );
   }
 
+  handleRangeChange = (date, view) => {
+    let { onRangeChange } = this.props
+    if(onRangeChange) {
+      let range = VIEWS[view].range(date, {})
```
[the remark this answers is not in the document]
Comment (jquense) [reply src/Calendar.js]: Also the console.warn is fine here, but use the `warning` module (already included in the package.json) https://www.npmjs.com/package/warning

File: src/Calendar.js [hunk 2]
```diff
@@ -679,6 +685,19 @@ class Calendar extends React.Component {
...
+      if(Array.isArray(range)) {
+        onRangeChange({ start: range[0], end: range[range.length -1]})
+      }
+      else{
+        onRangeChange(range)
```
Comment (jquense) [inline src/Calendar.js]: hmm I think if it's an array we want to pass all the values through, in the case of TimeGrig it maybe that there are "holes" in the range which is important to know. I'd just pass whatever the value is to `onRangeChange` and note in the doc comment that it's either an array or object for BULTIN views and YOLO for custom ones

**Comment by jquense** [comment]
@jhockett just a few small tweaks, thanks again for doing this!

**Comment by drewandrew** [comment]
Looking forward to having this option! Thanks @jhockett

**Comment by drewandrew** [comment]
@jhockett @jquense Before merging this feature please review the [PR referenced above](https://github.com/jhockett/react-big-calendar/pull/1) that fixes a bug in jhockett/master.

**Comment by wasabigeek** [comment]
I'm late to the discussion, but what's the advantage of onRangeChange vs directly using the `firstVisibleDay` and `lastVisibleDay` from [utils](https://github.com/intljusticemission/react-big-calendar/blob/master/src/utils/dates.js#L23)?

**Comment by drewandrew** [comment]
@wasabigeek I hadn't seen those util functions before, so thanks! If this PR stalls I'll likely fallback to using those.

See some extra discussion over at [this PR](https://github.com/jhockett/react-big-calendar/pull/1).

@jhockett I didn't realize that the first iteration of this PR used the `VIEWS` array. If support for custom views is the limiting factor (and I'm new to react-big-calendar as well so I'm unsure what, if any, of the other handler methods are broken for custom views), I'd point out that even the approved changes would be broken for custom views that don't have a `.range()` method in addition to being broken on first view changes.

This is out of the scope of this PR, but could the `VIEWS` array be generated dynamically? If not, I don't see a way for the `onRangeChange` method to work without breaking custom views.

**Comment by jhockett** [comment]
Thanks again @drewandrew for finding this bug. onRangeChange should now fire correctly for view changes

**Comment by drewandrew** [comment]
Aaaah good call @jhockett  :+1:

**Comment by drewandrew** [comment]
@jquense know you're busy with maintaining this awesome repo, but when you got a minute, please review this PR! :) Thanks

**Comment by jquense** [comment]
Can someone rebase and fix the conflicts?

**Pull request merged**